### PR TITLE
Fix handling of escaped characters in fstab

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/shared/generator.c [hunk 1]
```diff
@@ -216,8 +216,11 @@ int generator_write_timeouts(
         r = fstab_filter_options(opts, "comment=systemd.device-timeout\0"
                                        "x-systemd.device-timeout\0",
                                  NULL, &timeout, filtered);
-        if (r <= 0)
-                return r;
+        if (r <= 0) {
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse fstab options, ignoring: %m");
+                return 0;
+        }
```
Comment (poettering) [inline src/shared/generator.c:222]: doesn't matter at all, but maybe:

```c
if (r < 0) {
       log_warning_errno(…);
      return 0;
}
if (r == 0)
        return 0;
```

i.e. I like shallow nesting

File: src/basic/extract-word.h [hunk 2]
```diff
@@ -4,13 +4,15 @@
...
 
 typedef enum ExtractFlags {
-        EXTRACT_RELAX                    = 1 << 0,
-        EXTRACT_CUNESCAPE                = 1 << 1,
-        EXTRACT_CUNESCAPE_RELAX          = 1 << 2,
-        EXTRACT_UNESCAPE_SEPARATORS      = 1 << 3,
-        EXTRACT_UNQUOTE                  = 1 << 4,
-        EXTRACT_DONT_COALESCE_SEPARATORS = 1 << 5,
-        EXTRACT_RETAIN_ESCAPE            = 1 << 6,
+        EXTRACT_RELAX                    = 1 << 0, /* Allow unbalanced quote */
```
Comment (poettering) [inline src/basic/extract-word.h]: this also allows trailing escapes, but eats them up (unlike EXTRACT_CUNESCAPE_RELAX which allows them but maps them to themselves)

File: src/shared/fstab-util.c [hunk 3]
```diff
@@ -97,16 +97,17 @@ int fstab_filter_options(const char *opts, const char *names,
...
-                                size_t n = strcspn(end, ",");
+                                end += strcspn(end, ",\\");
 
-                                end += n;
-                                if (n > 0 && end[-1] == '\\')
-                                        end++;
-                                else
+                                if (IN_SET(*end, ',', '\0'))
                                         break;
+                                end ++;                 /* Skip the backslash */
```
Comment (poettering) [inline src/shared/fstab-util.c:168]: maybe encode the comment in code:

```c
assert(*end == '\\');
```

File: src/shared/fstab-util.c [hunk 4]
```diff
@@ -97,16 +97,17 @@ int fstab_filter_options(const char *opts, const char *names,
...
+                                if (*end != '\0')
+                                        end ++;         /* Skip the escaped char, but watch out for a trailing commma */
                         }
```
Comment (poettering) [inline src/shared/fstab-util.c:171]: can't we use extract_first_word() for this loop?

**Review by poettering — CHANGES_REQUESTED** [review]
looks good. minor stuff

**Review by anitazha — COMMENTED** [review]
Thanks for looking into this, however it doesn't fix #18952. Primarily because `x-systemd.requires` options are parsed using `fstab_extract_values` and not `fstab_filter_options` (why are there two anyways.....).

I put up https://github.com/systemd/systemd/pull/18960 with a test for `fstab_extract_values` but I'm not sure if this is the right way to go. Maybe it can be incorporated into this PR?

File: src/shared/generator.c (hunk 1 above)
Comment (keszybz) [reply src/shared/generator.c:222]: OK.

File: src/shared/fstab-util.c (hunk 4 above)
Comment (keszybz) [reply src/shared/fstab-util.c:171]: No. In this code path, failure is not allowed, so no allocations can be done.

**Comment by keszybz** [comment]
[Updated with the suggested changes.
Two new commits:
* shared/fstab-util: teach fstab_filter_options() a mode where all values are returned
* fstab-generator: get rid of fstab_extract_values()

This actually fixes the bug that @anitazha reported.

**Comment by keszybz** [comment]
@poettering liked this. It was changed a bit, but not much, and there are unit tests for the changed parts and the bug seems fixed. I'll merge this without additional review because I want this to be in -rc3.

File: src/basic/extract-word.h [hunk 5]
```diff
@@ -4,13 +4,15 @@
...
 
 typedef enum ExtractFlags {
-        EXTRACT_RELAX                    = 1 << 0,
-        EXTRACT_CUNESCAPE                = 1 << 1,
-        EXTRACT_CUNESCAPE_RELAX          = 1 << 2,
-        EXTRACT_UNESCAPE_SEPARATORS      = 1 << 3,
-        EXTRACT_UNQUOTE                  = 1 << 4,
-        EXTRACT_DONT_COALESCE_SEPARATORS = 1 << 5,
-        EXTRACT_RETAIN_ESCAPE            = 1 << 6,
+        EXTRACT_RELAX                    = 1 << 0, /* Allow unbalanced quote and eat up trailing backslash. */
```
Comment (bluca) [inline src/basic/extract-word.h:7]: thanks for this - I always had to go and try to figure out the code every time I needed to use one of these :-)

File: src/shared/fstab-util.c [hunk 6]
```diff
@@ -79,21 +79,80 @@ int fstab_is_mount_point(const char *mount) {
...
+                const char *names,
+                const char **ret_namefound,
+                char **ret_value,
+                char ***ret_values,
+                char **ret_filtered) {
+
         const char *name, *namefound = NULL, *x;
-        _cleanup_strv_free_ char **stor = NULL;
-        _cleanup_free_ char *v = NULL, **strv = NULL;
+        _cleanup_strv_free_ char **stor = NULL, **values = NULL;
```
Comment (poettering) [inline src/shared/fstab-util.c:91]: `stor` can be be moved into the inner block

File: src/shared/fstab-util.c [hunk 7]
```diff
@@ -79,21 +79,80 @@ int fstab_is_mount_point(const char *mount) {
...
+         * Returns negative on error, true if any matching options were found, false otherwise. */
+
+        if (ret_filtered || ret_value || ret_values) {
+                /* For backwards compatibility, we need to pass-through escape characters.
+                 * The only ones we "consume" are the ones used as "\," or "\\". */
+                r = strv_split_full(&stor, opts, ",", EXTRACT_UNESCAPE_SEPARATORS | EXTRACT_UNESCAPE_RELAX);
+                if (r < 0)
+                        return r;
+
+                filtered = memdup(stor, sizeof(char*) * (strv_length(stor) + 1));
```
Comment (poettering) [inline src/shared/fstab-util.c:118]: `newdup(char*, stor, strv_length(stor) + 1);`

**Review by poettering — COMMENTED** [review]
lgtm, just some minor comment

File: src/shared/fstab-util.c (hunk 6 above)
Comment (keszybz) [reply src/shared/fstab-util.c:91]: Actually it's used outside of the inner block. I'll add a comment.

File: src/shared/fstab-util.c (hunk 7 above)
Comment (keszybz) [reply src/shared/fstab-util.c:118]: Ack.

**Pull request merged**